[Homework #18] - React UI by Pavel Didenko

Win4esterD opened the pull request:
@ilya-ischenko Solved the task, seems like the app works fine.

ilya-ischenko commented on homeworks/pavel-didenko/18-react-ui/index.jsx:
```diff
@@ -0,0 +1,78 @@
+const App = () => {
+  const [allUsers, setAllUsers] = useState([]);
+  const [users, setUsers] = useState([]);
+  const url = 'https://tinyfac.es/api/data?limit=50&quality=0';
```
It can be moved outside the component and exported in order to made it reusable

ilya-ischenko commented on homeworks/pavel-didenko/18-react-ui/components/refreshUsers/RefreshAllUsers.jsx:
```diff
@@ -0,0 +1,24 @@
+const RefreshAllUsers = ({setUsers, fetchUsers, users, setAllUsers}) => {
+  async function refreshAll() {
+    const newUsers = await fetchUsers();
+    setAllUsers(newUsers);
+    setUsers(users => {
+      return users.map((_, index) => {
+        if (index < newUsers.length) {
+          return newUsers[index].url;
+        }
+      });
+    });
+    
+  }
+
+  return (
+    <button
+      className="refresh-button"
+      onClick={refreshAll}
+      style={{marginTop: users.length < 10 ? '20.9rem' : '3rem'}}
+    >
+      Refresh All
+    </button>
+  );
+};
```
Bad one. This should be just ui component, and it should not be responsible for updating users.
Try to keep logic in one place and not to spread it across your components.
So there you just need to pass onClick prop and use it like this - " onClick={onClick}", all logic should be in the parent component in this case.

Win4esterD replied:
Moved the logic outside the component.

ilya-ischenko commented on homeworks/pavel-didenko/18-react-ui/components/user/Users.jsx:
```diff
@@ -0,0 +1,25 @@
+const User = ({link, index, setUsers, fetchUsers}) => {
+
+  async function updateUserAvatar() {
+    const [newAvatar] = await fetchUsers();
+    setUsers(users =>
+      users.map((user, i) => {
+        if (i === index) {
+          return newAvatar.url;
+        }
+
+        return user;
+      }),
+    );
+  }
+  return (
+    <div className="user">
+      <img className="user__foto" src={link} alt="user image" />
+      <img
+        className="user__refresh-foto"
+        src="./assets/img/icons/update-user.svg"
+        onClick={updateUserAvatar}
+      />
+    </div>
+  );
+};
```
1. The same as for RefreshAllUsers 
2. Bad file name, should be User.jsx, not Users.

Win4esterD replied:
1) Moved logic outside the component.
2) Renamed to User.jsx

ilya-ischenko reviewed (commented):
1. In my case refresh button outside visible area, so I should scroll down to see it. Did you plan so?
![Screenshot 2023-09-26 at 12 14 29](https://github.com/qaprosoft/react-laba-international-2/assets/97586975/888735ea-e93a-4f04-b34d-b016eb97e830)
2. Why do not you use api to get random user? All logic should be on back-end side, so you just need to fetch users one by one and not fetch 50 users once and create logic on front-end side, it should be back end responsibility. 
3. Feel free to use crete-react or next app, or vite to initialize react project. It will be more comfortable to work with react project.
4. Also you can add "loading state" with loading animation for each tile

Win4esterD commented:
@ilya-ischenko 

1) Not that I planed this, bit it was designer's issue, I just made it pixel perfect, haha
![planed_design](https://github.com/qaprosoft/react-laba-international-2/assets/85812541/383277d2-a2a7-44cc-b549-b4f8167a8587)
For now I decreased idents, so it's resolved.

2) I thought that API sends pictures in the same order, so I implemented getRandomUser and then didn't remove it. Now it's removed.

3) OK, will do, didn't know that it was allowed. Saw you had written "You can use create React App since next task". For me the next one was 19. :)

4) I implemented animation using other approach. When you click on a single tile, it runs animation without changing state. If you click on "Refresh All", I use "loading" state to trigger animation for all tiles. Hope it's fine.

ilya-ischenko commented on homeworks/pavel-didenko/18-react-ui/index.jsx:
```diff
@@ -0,0 +1,77 @@
+  return (
+    <main>
+      <div className="image-wrapper">
+        {users.map((user, index) => (
+          <User
+            link={user}
+            key={index}
+            index={index}
+            updateUserAvatar={updateUserAvatar}
+            loading={loading}
+          />
```
"event.target.getAttribute('index')", is not so optimal solution.
You can do it in this way:   updateUserAvatar={() => updateUserAvatar(index)}

Win4esterD replied:
I don't understand how to implement this solution. Since I store fotos urls in the "users" state, maybe it would be fine to get event.target.src and update state and DOM using it? My solution for now looks like this:

![image](https://github.com/qaprosoft/react-laba-international-2/assets/85812541/2894d280-fc6d-4660-9eaa-3dfbbbd54233)

ilya-ischenko reviewed (commented):
1. Always use CamelCase for components name, incl. component's folder. So rename folder user to User, refreshUsers to **RefreshAllUsersButton**.
2. I suppose api url should be in "constants" folder, not modules, modules sounds very similar to packages
3. Pay attention to UX. Hover and click area should be the same, because as user I expect that I can click on this card if I am able to see "refresh icon", but not in your case.
![Screenshot 2023-09-27 at 13 57 14](https://github.com/qaprosoft/react-laba-international-2/assets/97586975/064a0fc1-dc6a-48d6-8ef8-4e9ae6dc9767)
 case, this is not convenient

Win4esterD commented:
@ilya-ischenko

1) Renamed component to "RefreshAllUsersButton".
2) Renamed "modules" to "constants"
3) Fixed UI, now it triggers update when you either click on an avatar or on update image.

ilya-ischenko reviewed (approved):
Well done!
Approved.
